main/wallet.js

```diff
@@ -0,0 +1,435 @@
+'use strict'
+
+const keytar = require('keytar')
+const { generateMnemonic } = require('@zondax/filecoin-signing-tools')
+const { default: Filecoin, HDWalletProvider } = require('@glif/filecoin-wallet-provider')
+const { CoinType } = require('@glif/filecoin-address')
+const electronLog = require('electron-log')
+const assert = require('assert')
+const { request, gql } = require('graphql-request')
+const { FilecoinNumber, BigNumber } = require('@glif/filecoin-number')
+const { Message } = require('@glif/filecoin-message')
+const { getDestinationWalletAddress } = require('./station-config')
+const timers = require('node:timers/promises')
+const Store = require('electron-store')
+
+/** @typedef {import('./typings').GQLMessage} GQLMessage */
+/** @typedef {import('./typings').GQLStateReplay} GQLStateReplay */
+/** @typedef {import('./typings').GQLTipset} GQLTipset */
+/** @typedef {import('./typings').Context} Context */
+/** @typedef {import('./typings').FILTransaction} FILTransaction */
+/** @typedef {import('./typings').FILTransactionProcessing} FILTransactionProcessing */
+/** @typedef {import('./typings').FILTransactionLoading} FILTransactionLoading */
+/** @typedef {import('./typings').TransactionStatus} TransactionStatus */
+
+const log = electronLog.scope('wallet')
+const url = 'https://graph.glif.link/query'
+const walletStore = new Store({
+  name: 'wallet'
+})
+
+let address = ''
+/** @type {Filecoin | null} */
+let provider = null
+/** @type {Context | null} */
+let ctx = null
+let transactions = loadStoredEntries()
+let balance = loadBalance()
+/** @type {Set<string>} */
+const stateReplaysBeingFetched = new Set()
+
+/**
+ * @returns {Promise<string>}
+ */
+async function getSeedPhrase () {
+  const service = 'filecoin-station-wallet'
+  let seed = await keytar.getPassword(service, 'seed')
+  if (seed) {
+    log.info('Using existing seed phrase')
+  } else {
+    seed = generateMnemonic()
+    await keytar.setPassword(service, 'seed', seed)
+    log.info('Created new seed phrase')
+  }
+  return seed
+}
+
+/**
+ * @param {Context} _ctx
+ */
+async function setup (_ctx) {
+  ctx = _ctx
+
+  const seed = await getSeedPhrase()
+  provider = new Filecoin(new HDWalletProvider(seed), {
+    apiAddress: 'https://api.node.glif.io/rpc/v0'
+  })
+  ;[address] = await provider.wallet.getAccounts(
+    0,
+    1,
+    CoinType.MAIN
+  )
+  log.info('Address: %s', address)
+
+  ;(async () => {
```

Comment:
I've refactored this to add the `updateWalletData` function, I think that's an improvement, but I don't see it the same way about `setInterval`. Now that we have `async/await` available, `setInterval` isn't necessary any more and can be forgotten when writing any new code. `setInterval` is a platform feature, while you can do the same thing just with language features too. I think `setInterval` was invented because we didn't have `async/await` yet, but now it can be decomissioned.

Reply:
> BTW should we fetch the balance & the list of transactions concurrently?

I think it's better UX not to do it, so that the balance always changes before the transactions, otherwise you might think a transaction didn't have any effect. Ideally we always show a consistent state, but that's not possible with these APIs

Reply:
> I think it's better UX not to do it, so that the balance always changes before the transactions

Makes sense 👍🏻

Reply:
TL;DR: **I am fine to keep your current version. Extracting `refreshState` helped a lot 👍🏻**

> I've refactored this to add the `updateWalletData` function, I think that's an improvement, but I don't see it the same way about `setInterval`. Now that we have `async/await` available, `setInterval` isn't necessary any more and can be forgotten when writing any new code. `setInterval` is a platform feature, while you can do the same thing just with language features too. I think `setInterval` was invented because we didn't have `async/await` yet, but now it can be decommissioned.

Well, you cannot avoid platform APIs entirely-  you still need `timers.setTimeout`.

```js
// I like this more:
setInterval(doSomeWork, 10_000)

// Rather than
;(async () => {
  while (true) { 
    await doSomeWork()
    await timers.setTimeout(10_000)
  }
})()
```

Maybe you are missing a helper function that would allow us to write something like this?

```js
repeatedlyRunInBackground(refreshState, 10_000)

// implementation
function repeatedlyRunInBackground(taskFn, intervalInMs) {
  ;(async () => {
    while (true) { 
      await taskFn()
      await timers.setTimeout(intervalInMs)
    }
  })()
}
```

Anyhow. This is a relatively minor detail. The code is isolated and easy to rework later if we ever change our mind.

I am fine to keep your current version. Extracting `refreshState` helped a lot 👍🏻